contracts/Staking.sol

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.6.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@nomiclabs/buidler/console.sol";
+
+contract Staking {
+    // timestamp for the epoch 1
+    // everything before that is considered epoch 0 which won't have a reward but allows for the initial stake
+    uint256 public epoch1Start;
+
+    // duration of each epoch
+    uint256 epochDuration;
+
+    mapping(address => mapping(address => uint256)) private balances;
+
+    struct Pool {
+        uint256 size;
+        bool set;
+    }
+
+    // for each token, we store the total pool size
+    mapping(address => mapping(uint256 => Pool)) private poolSize;
+
+    struct Checkpoint {
+        uint256 epochId;
+        uint256 balance;
+    }
+
+    // balanceCheckpoints[user][token][]
+    mapping(address => mapping(address => Checkpoint[])) private balanceCheckpoints;
+
+    /*
+     *
+     */
+    constructor (uint256 _epoch1Start, uint256 _epochDuration) public {
+        epoch1Start = _epoch1Start;
+        epochDuration = _epochDuration;
+    }
+
+    /*
+     * Stores `amount` of `tokenAddress` tokens for the `user` into the vault
+     */
+    function deposit(address tokenAddress, uint256 amount) public {
+        require(amount > 0, "Staking: Amount must be > 0");
+
+        IERC20 token = IERC20(tokenAddress);
+        uint256 allowance = token.allowance(msg.sender, address(this));
+        require(allowance >= amount, "Staking: Token allowance too small");
+
+        balances[msg.sender][tokenAddress] += amount;
+
+        token.transferFrom(msg.sender, address(this), amount);
```

Comment:
I believe transferFrom should be above balances+= amount. Also please use safemath.

Reply:
It was a bug some time ago related to this.

Reply:
Usually that's how I would have done it but in this case I was aiming for the checks-effects-interactions pattern which avoids re-entrancy attacks. I can change it if there was a different kind of bug affecting it.

Reply:
just to be sure we could use OpenZeppelin ReentrancyGuard and add the modifier to this method, this way you could follow the current way to do it.

Reply:
done!